### ISSUE TITLE
Memory allocation issue in Base64Util.encodeBase64 

### ISSUE THREAD

**Issue by leonardolombardo** [nextgenhealthcare/connect#4392]
![image](https://user-images.githubusercontent.com/6245101/83253889-0ea07680-a1ae-11ea-829d-c2e2c493d50f.png)

In class com.mirth.connect.donkey.util.Base64Util method encodeBase64 there is an allocation of memory that could be massive and lead to an OutOfMemory (see image).

I think it'd be better to handle this case putting a cap to the size passed to ByteArrayOutputStream constructor.

**Comment by ChristopherSchultz** [comment]
Base64 encoding generates 4 bytes of output for every 3 bytes of input. Under what conditions would this code cause a `ByteArrayOutputStream` that was initially-created with a size causing an OOME but not cause an OOME once it was completely-filled with data?

If you use a standard `ByteArrayOutputStream` without pre-allocating it, it will probably grow too large and will be more likely to cause OOME.

The only possible protection I can see here would be to allocate exactly `Math.ceil(bytes.length/3) >> 2` bytes instead of multiplying by 1.4, which is slightly sloppy.

Pre-allocation is better, and the exact size *can* be computed. The buffer *will* be filled, and if it's going to cause an OOME, it doesn't matter whether it happens on line 50 or line 59 of your screenshot.

**Comment by leonardolombardo** [comment]
Sorry, you're completely right.

This requires a little bit more investigation. I'm getting OOME when doing addAttachment with large files. Does Connect supports a stream-based attachment operation that I missed ?

**Comment by ChristopherSchultz** [comment]
I don't believe MC provides any stream.based conversion. All of the puzzle pieces are there, however, to convert to base64 encoding in a more streaming kind of way. I have no idea what is available for "attachments" in Mirth, though... it may be that you have to attach the whole object all at once, so perhaps large attachments in Mirth are always a problem.

I'm speaking from a position of (significant) ignorance, here, though. I would ask on the Slack channel about streaming attachments.
